### ISSUE TITLE
Insufficient narrowing with temporary values

### ISSUE THREAD

**Issue by vberlier** [microsoft/pyright#2831]
Pyright often reports errors where it shouldn't when passing temporary values as arguments.

```python
FileChanges = Dict[str, Literal["created", "edited", "removed"]]

changes: FileChanges = {}
changes.update({filename: "removed" for filename in ["foo.py", "bar.py"]})
```

```
Argument of type "dict[str, str]" cannot be assigned to parameter "__m" of type "Iterable[tuple[str, Literal['created', 'edited', 'removed']]]"
```

Since the result of the comprehension can't be used anywhere else it should be ok to narrow it to the desired type.

**Comment by erictraut** [comment]
Thanks for the bug report. 

The TypeVar constraint solver in pyright contains various heuristics to determine whether a TypeVar can be bound to a literal value or whether the literal part of the type should be "stripped" and the corresponding object type (e.g. `str`) should be bound instead. The existing heuristics failed in this case. I've added more logic to fix this hole.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.206, which I just published. It will also be in the next release of pylance.
